File: backend/workflow/discover.go
package workflow

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type scriptCandidate struct {
	path    string
	modTime int64
}

// FindLatestScript returns the most recently modified workflow.js under workDir/.enough/workflows/.
func FindLatestScript(workDir string) (string, error) {
	root := filepath.Join(workDir, ".enough", "workflows")
	matches, err := filepath.Glob(filepath.Join(root, "*", "workflow.js"))
	if err != nil {
		return "", err
	}
	var candidates []scriptCandidate
	for _, path := range matches {
		info, statErr := os.Stat(path)
		if statErr != nil || info.IsDir() {
			continue
		}
		candidates = append(candidates, scriptCandidate{path: path, modTime: info.ModTime().UnixNano()})
	}
	if len(candidates) == 0 {
		return "", fmt.Errorf("no workflow scripts found under %s", root)
	}
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].modTime > candidates[j].modTime
	})
	return candidates[0].path, nil
}

// ResolveScriptPath resolves a workflow script path relative to workDir when needed.
func ResolveScriptPath(workDir, arg string) (string, error) {
	arg = strings.TrimSpace(arg)
	if arg == "" {
		return "", fmt.Errorf("empty workflow path")
	}
	path := arg
	if !filepath.IsAbs(path) {
		path = filepath.Join(workDir, path)
	}
	path = filepath.Clean(path)
	info, err := os.Stat(path)
	if err != nil {
		return "", err
	}
	if info.IsDir() {
		path = filepath.Join(path, "workflow.js")
		info, err = os.Stat(path)
		if err != nil {
			return "", err
		}
	}
	if info.IsDir() {
		return "", fmt.Errorf("workflow path is a directory: %s", path)
	}
	return path, nil
}

File: backend/workflow/discover.ts
// PORT: backend/workflow/discover.go

import { Effect } from "effect";
import fs from "node:fs";
import path from "node:path";

interface ScriptCandidate {
  path: string;
  modTime: number;
}

export function FindLatestScript(workDir: string): Effect.Effect<string, Error> {
  return Effect.try({
    try: () => {
      const root = path.join(workDir, ".hollow", "workflows");
      if (!fs.existsSync(root)) {
        throw new Error(`no workflow scripts found under ${root}`);
      }

      const entries = fs.readdirSync(root, { withFileTypes: true });
      const candidates: ScriptCandidate[] = [];

      for (const entry of entries) {
        if (entry.isDirectory()) {
          const wfPath = path.join(root, entry.name, "workflow.js");
          try {
            const stat = fs.statSync(wfPath);
            if (stat.isFile()) {
              candidates.push({
                path: wfPath,
                modTime: stat.mtime.getTime(),
              });
            }
          } catch {}
        }
      }

      if (candidates.length === 0) {
        throw new Error(`no workflow scripts found under ${root}`);
      }

      candidates.sort((a, b) => b.modTime - a.modTime);
      return candidates[0].path;
    },
    catch: (cause) => (cause instanceof Error ? cause : new Error(String(cause))),
  });
}

export function ResolveScriptPath(workDir: string, arg: string): Effect.Effect<string, Error> {
  return Effect.try({
    try: () => {
      const trimmed = arg.trim();
      if (trimmed === "") {
        throw new Error("empty workflow path");
      }

      let scriptPath = trimmed;
      if (!path.isAbsolute(scriptPath)) {
        scriptPath = path.join(workDir, scriptPath);
      }
      scriptPath = path.normalize(scriptPath);

      let stat = fs.statSync(scriptPath);
      if (stat.isDirectory()) {
        scriptPath = path.join(scriptPath, "workflow.js");
        stat = fs.statSync(scriptPath);
      }

      if (stat.isDirectory()) {
        throw new Error(`workflow path is a directory: ${scriptPath}`);
      }

      return scriptPath;
    },
    catch: (cause) => (cause instanceof Error ? cause : new Error(String(cause))),
  });
}

/*
PORT STATUS
source path: backend/workflow/discover.go
source lines: 67
draft lines: 83
confidence: high
status: phase_b_compile
*/
